scanner/src/main/kotlin/experimental/NestedProvenanceResolver.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2021 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.IOException
+
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.model.ArtifactProvenance
+import org.ossreviewtoolkit.model.KnownProvenance
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.utils.createOrtTempDir
+
+/**
+ * The [NestedProvenanceResolver] provides a function to resolve nested provenances.
+ */
+interface NestedProvenanceResolver {
+    /**
+     * Resolve nested [Provenance]s of the provided [provenance]. For an [ArtifactProvenance] the returned
+     * [NestedProvenance] always contains only the provided [ArtifactProvenance]. For a [RepositoryProvenance] the
+     * resolver looks for nested repositories, for example Git submodules or Mercurial subrepositories.
+     */
+    fun resolveNestedProvenance(provenance: KnownProvenance): NestedProvenance
+}
+
+/**
+ * The default implementation of [NestedProvenanceResolver].
+ */
+class DefaultNestedProvenanceResolver : NestedProvenanceResolver {
+    override fun resolveNestedProvenance(provenance: KnownProvenance): NestedProvenance {
+        return when (provenance) {
+            is ArtifactProvenance -> NestedProvenance(root = provenance, subRepositories = emptyMap())
+            is RepositoryProvenance -> resolveNestedRepository(provenance)
+        }
+    }
+
+    private fun resolveNestedRepository(provenance: RepositoryProvenance): NestedProvenance {
+        val vcs = VersionControlSystem.forType(provenance.vcsInfo.type)
+            ?: VersionControlSystem.forUrl(provenance.vcsInfo.url)
+            ?: throw IOException("Could not determine VCS type for ${provenance.vcsInfo}.")
+
+        val targetDir = createOrtTempDir()
```

Comment:
Looks like this call again causes

```
error: Cannot fetch oss-review-toolkit/ort-test-data-git-submodules\test-data-npm\isarray (GitError: oss-review-toolkit/ort-test-data-git-submodules\test-data-npm\isarray init: C:/Users/VssAdministrator/AppData/Local/Temp/ort-DefaultNestedProvenanceResolver4660852545205957772/.repo/projects/submodules.git/subprojects/test-data-npm.git/subproject-objects/oss-review-toolkit/ort-test-data-git-submodules/test-data-npm/isarray.git: Filename too long
```

on Windows. If you're lucky, you can work around it like in

https://github.com/oss-review-toolkit/ort/blob/a754d7bfc2e4687a884d62d0f516a23c83b3e38a/analyzer/src/funTest/kotlin/GitRepoFunTest.kt#L52-L54

Reply:
Thanks, for `Git` we have:
https://github.com/oss-review-toolkit/ort/blob/a754d7bfc2e4687a884d62d0f516a23c83b3e38a/downloader/src/main/kotlin/vcs/Git.kt#L144-L146

Wouldn't it be an option to do the same for `GitRepo`?

Reply:
I tried doing this in https://github.com/oss-review-toolkit/ort/pull/4024, but as `repo` is written in Python, somewhere in the JVM -> ORT - Python -> repo process chain not all environment variables seem to get passes on as expected, and also another approach via setting / restoring `git config` did not succeed initially, and then I ran out of time to try further.

Reply:
Ok, I decided to disable the one failing test on Windows and leave the problem for later, because the workaround didn't work.